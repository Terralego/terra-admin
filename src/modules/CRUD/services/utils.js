import Api from '@terralego/core/modules/Api';
import { geomTypes } from '../../RA/DataSource/index';


import {
  POINT,
  LINESTRING,
  POLYGON,
  MULTI_POINT,
  MULTI_LINESTRING,
  MULTI_POLYGON,
  GEOMETRY_COLLECTION,
} from '../../../utils/geom';

export const getObjectOrderedValue = (objectValues, arrayOrder = []) => {
  if (!objectValues) {
    return {};
  }
  const UIOrderReverse = [...arrayOrder].reverse();
  return Object.keys(objectValues).sort(
    ((a, b) => UIOrderReverse.indexOf(b) - UIOrderReverse.indexOf(a)),
  ).reduce((acc, prop) => (
    { ...acc, [prop]: objectValues[prop] }
  ), {});
};

export const isTableObject = (arrayOfObjects = []) => {
  if (!arrayOfObjects.length) {
    return false;
  }
  const firstKeys = Object.keys(arrayOfObjects[0]).join();
  return arrayOfObjects.every(obj => Object.keys(obj).join() === firstKeys);
};

export const sanitizeCustomEndpoint = str => {
  if (str.startsWith('/api/')) {
    return str.replace('/api/', '');
  }
  return str;
};

export const exportFileFromURL = async (url, name) => {
  const data = await Api.request(sanitizeCustomEndpoint(url), { responseType: 'text' });
  const file = new Blob([data], { type: 'text/plain' });

  const [fileName] = url.split('/').slice(-1);
  const link = document.createElement('a');
  link.download = name || fileName;
  link.href = global.URL.createObjectURL(file);
  link.click();
};


const getSchemaCoordinates = coordinates => {
  if (!Array.isArray(coordinates[0])) {
    return { type: 'number' };
  }
  return {
    items: getSchemaCoordinates(coordinates[0]),
    type: 'array',
  };
};

const getUISchemaCoordinates = coordinates => {
  const uiOptions = {
    addable: false,
    orderable: false,
    removable: false,
  };
  if (!Array.isArray(coordinates[0])) {
    return {
      'ui:options': uiOptions,
    };
  }
  return {
    items: getUISchemaCoordinates(coordinates[0]),
    'ui:options': uiOptions,
  };
};

const getDefaultCoordinatesByGeomType = geomType => {
  switch (geomType) {
    case LINESTRING:
    case MULTI_POINT:
      return [[], []];
    case POLYGON:
    case MULTI_LINESTRING:
      return [[[], []]];
    case MULTI_POLYGON:
      return [[[[], []]]];
    case GEOMETRY_COLLECTION:
      return [{}];
    case POINT:
    default:
      return [];
  }
};

export const getJSONSchemaFromGeom = (
  { identifier, geom, geom_type: geomType, title, ...rest },
  notRequired,
) => {
  const type = geomTypes[geomType];
  const value = geom || { type, coordinates: [] };
  const { coordinates } = geom || {};
  const coordinatesCalculated = coordinates && coordinates.length
    ? coordinates
    : getDefaultCoordinatesByGeomType(geomType);
  return {
    ...rest,
    display_value: value,
    method: (identifier !== null) ? 'PATCH' : 'POST',
    schema: {
      properties: {
        type: {
          type: 'string',
          title: 'Type',
        },
        coordinates: {
          items: getSchemaCoordinates(coordinatesCalculated),
          type: 'array',
        },
      },
      required: ['type', !notRequired && 'coordinates'].filter(Boolean),
      title,
      type: 'object',
    },
    ui_schema: {
      'ui:field': 'geometry',
      coordinates: getUISchemaCoordinates(coordinatesCalculated),
      type: {
        'ui:widget': 'hidden',
      },
    },
    title,
    type: 'object',
    value,
  };
};

/**
 * Clean all useless properties and keep required properties
 * 1/ Flat all props by removing groups
 * 2/ Keep only required props
 *
 * @param {Object} props JSON schema properties
 * @returns {Object} JSON schema properties cleaned
 */
export const requiredProperties = props => {
  const propertiesWithoutGroup = Object.values(props).reduce((list, value) => {
    const { type, properties, required = [] } = value;

    if (type === 'object' && required.length) {
      const requiredProps = Object.entries(properties).reduce((acc, [subKey, subValue]) => (
        required.includes(subKey)
          ? ({ ...acc, [subKey]: subValue })
          : acc
      ), {});

      return {
        ...list,
        properties: { ...list.properties, ...requiredProps },
        required: [...list.required, ...required],
      };
    }

    return list;
  }, { properties: {}, required: [] });

  return propertiesWithoutGroup;
};

export default {
  exportFileFromURL,
  getJSONSchemaFromGeom,
  getObjectOrderedValue,
  isTableObject,
  requiredProperties,
  sanitizeCustomEndpoint,
};
