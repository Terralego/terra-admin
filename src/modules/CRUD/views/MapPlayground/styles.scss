.CRUD-no-settings {
  margin: auto;
  width: max-content;
  max-width: 50%;
  min-width: 200px;
}

.CRUD-map {
  flex-grow: 1;

  .interactive-map {
    position: absolute;
    top: 0;
    right: 0;
    left: 0;
    height: 100%;
    z-index: 1;
  }

  &--with-table {
    &-medium {
      .interactive-map {
        .mapboxgl-ctrl-bottom-left,
        .mapboxgl-ctrl-bottom-right {
          bottom: calc(100% / 3);
        }
      }
    }
    &-minified {
      .interactive-map {
        .mapboxgl-ctrl-bottom-left,
        .mapboxgl-ctrl-bottom-right {
          bottom: 3rem;
        }
      }
    }
  }

  .mapboxgl-ctrl-search {
    overflow: visible;
    .search-results__item button {
      height: auto;
    }
  }
}

.CRUD-no-map {
  position: absolute;
  top: 0;
}

.CRUD-table {
  position: relative;
  flex-shrink: 0;
  padding: 0;
  height: 0;
  transition: height 0.2s ease-in;
  display: flex;
  margin-top: auto;
  overflow: hidden;
  &--medium {
    height: calc(100% / 3);
  }
  &--full {
    height: 100%;
  }
  &--minified {
    height: 3rem;
    .table-header + *,
    .table-footer {
      display: none;
    }
  }

  .table-container {
    background: #fff;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 10;
    display: flex;
    flex-direction: column;

    .table {
      display: flex;
      flex-direction: column;
      overflow: auto;
    }
  }
}
