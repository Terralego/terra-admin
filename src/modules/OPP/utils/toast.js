import { Intent, Toaster } from '@blueprintjs/core';

const submitToaster = Toaster.create();

function displayToaster (viewpoint, translationSucess, translationError) {
  submitToaster.show({
    message: viewpoint.id ? translationSucess : translationError,
    intent: viewpoint.id ? Intent.SUCCESS : Intent.DANGER,
  });
}

function displayError (translationError) {
  submitToaster.show({
    message: translationError,
    intent: Intent.DANGER,
  });
}

export const toast = {
  submitToaster,
  displayToaster,
  displayError,
};

export default { toast };
