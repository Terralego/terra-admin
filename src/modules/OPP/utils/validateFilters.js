const isEmpty = value => value.length < 1
  || (Array.isArray(value) && !value.some(elem => elem !== null))
  || (typeof value === 'string' && value === '');

const removeEmptyProperties = obj =>
  Object.keys(obj).reduce((acc, key) =>
    (isEmpty(obj[key]) ? acc : { ...acc, [key]: obj[key] }), {});

export function parsePropertiesToData (properties) {
  const data = removeEmptyProperties(properties);
  let parseDate = {};

  if (data.viewpointDate) {
    if (data.viewpointDate.every(date => date !== null)) {
      parseDate = {
        date__form: data.viewpointDate[0].toLocaleDateString(),
        date__to: data.viewpointDate[1].toLocaleDateString(),
      };
    } else {
      const pos = data.viewpointDate.findIndex(date => date !== null);
      parseDate = pos === 0
        ? { date__form: data.viewpointDate[pos].toLocaleDateString() }
        : { date__to: data.viewpointDate[pos].toLocaleDateString() };
    }

    // Destructuring to remove useless properties viewpointDate
    const { viewpointDate, ...props } = data;
    return {
      ...props,
      ...parseDate,
    };
  }
  return data;
}

export const isDate = date => new Date(date).getDate() > 0;

export default { parsePropertiesToData, isDate };
