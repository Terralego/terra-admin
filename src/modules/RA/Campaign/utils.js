export const stateChoices = [
  { id: 'draft', name: 'resources.campaign.states.draft' },
  { id: 'started', name: 'resources.campaign.states.started' },
  { id: 'closed', name: 'resources.campaign.states.closed' },
];

export default {
  stateChoices,
};
