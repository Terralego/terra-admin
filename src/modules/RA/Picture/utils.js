export const stateChoices = [
  { id: 'draft', name: 'resources.picture.states.draft' },
  { id: 'submited', name: 'resources.picture.states.submited' },
  { id: 'accepted', name: 'resources.picture.states.accepted' },
  { id: 'refused', name: 'resources.picture.states.refused' },
];

export default {
  stateChoices,
};
