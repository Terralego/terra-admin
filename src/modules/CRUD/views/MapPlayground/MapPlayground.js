import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Redirect } from 'react-router-dom';

import Loading from '../../../../components/Loading';
import Message from '../../../../components/Message';
import DataTable from '../../components/DataTable';
import DetailsWrapper from '../../components/DetailsWrapper';
import Details from '../../components/Details';
import Map from './components/Map';

import { getFirstCrudViewName, getView } from '../../services/CRUD';
import { TABLE_MEDIUM } from '../../services/UserSettingsProvider';
import { generateURI } from '../../config';
import { toast } from '../../../../utils/toast';


import './styles.scss';

export class MapPlayground extends React.Component {
  static propTypes = {
    featureToHighlight: PropTypes.func,
    getSettings: PropTypes.func.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        id: PropTypes.string,
        layer: PropTypes.string,
      }),
    }),
    settings: PropTypes.shape({}),
    settingsEndpoint: PropTypes.string,
    tableSize: PropTypes.string,
  };

  static defaultProps = {
    featureToHighlight: () => {},
    match: {
      params: {
        id: undefined,
        layer: undefined,
      },
    },
    settings: undefined,
    settingsEndpoint: undefined,
    tableSize: TABLE_MEDIUM,
  }

  state = {
    tableTransitioned: TABLE_MEDIUM,
  }

  componentDidMount () {
    const {
      getSettings,
      settingsEndpoint,
      settings,
    } = this.props;

    if (!settings) {
      getSettings(settingsEndpoint);
    }
  }

  onTableHoverCell = (featureId, hover = true) => {
    const {
      featureToHighlight,
      match: { params: { layer } },
    } = this.props;
    featureToHighlight({ layer, featureId, hover });
  }

  onTableTransitionedEnd = event => {
    event.preventDefault();
    if (event.target !== event.currentTarget) {
      return;
    }

    const {
      match: { params: { id } },
      tableSize,
    } = this.props;

    this.setState(({ tableTransitioned: prevTableTransitioned }) => {
      const nextTableTransitioned = id ? 'hidden' : tableSize;
      if (prevTableTransitioned === nextTableTransitioned) {
        return null;
      }
      return {
        tableTransitioned: nextTableTransitioned,
      };
    });
  }

  render () {
    const { tableTransitioned } = this.state;

    const {
      dataTableRef,
      detailsRef,
      settings,
      match: { params: { layer, id } },
      t,
      tableSize,
      errors,
    } = this.props;

    if (errors.settings) {
      return (
        <Message intent="danger" className="CRUD-no-settings">
          {t('CRUD.settings.unableToLoad')}
        </Message>
      );
    }

    if (!settings) {
      return <Loading spinner />;
    }

    const firstCrudViewName = getFirstCrudViewName(settings);
    const redirectArgs = firstCrudViewName ? { layer: firstCrudViewName } : {};

    if (layer && !getView(settings, layer)) {
      toast.displayError(t('CRUD.layer.errorNoLayer'));
      return <Redirect to={generateURI('layer', redirectArgs)} />;
    }
    // Redirect to the first item of the menu when no one is selected
    if (layer === undefined && firstCrudViewName) {
      return <Redirect to={generateURI('layer', redirectArgs)} />;
    }

    const areDetailsVisible = !!id;

    return (
      <>
        <DetailsWrapper detailsRef={detailsRef}>
          {areDetailsVisible && (
            <Details />
          )}
        </DetailsWrapper>

        <div
          className={classnames(
            {
              'CRUD-map': true,
              [`CRUD-map--with-table-${tableSize}`]: layer && !areDetailsVisible,
            },
          )}
        >
          <Map
            triggerFitBound={tableTransitioned}
          />
        </div>

        <div
          ref={dataTableRef}
          onTransitionEnd={this.onTableTransitionedEnd}
          className={classnames(
            {
              'CRUD-table': true,
              'CRUD-table--active': layer && !areDetailsVisible,
              [`CRUD-table--${tableSize}`]: layer && !areDetailsVisible,
            },
          )}
        >
          <DataTable
            layerName={layer}
            onHoverCell={this.onTableHoverCell}
          />
        </div>
      </>
    );
  }
}

export default MapPlayground;
