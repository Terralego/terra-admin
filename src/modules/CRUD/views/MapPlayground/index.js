import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';

import { connectAppProvider } from '../../../../components/AppProvider';
import { connectCRUDProvider } from '../../services/CRUDProvider';
import { withTableSize } from '../../services/UserSettingsProvider';
import { connectMapProvider } from '../../services/MapProvider';
import compose from '../../../../utils/compose';

import MapPlayground from './MapPlayground';

const appProviderGetter = ({
  env: {
    modules: { CRUD: { settings } },
  },
}) => ({
  settingsEndpoint: settings,
});


export default compose(
  withRouter,
  connectAppProvider(appProviderGetter),
  connectCRUDProvider('errors', 'getSettings', 'settings'),
  connectMapProvider('dataTableRef', 'detailsRef', 'featureToHighlight'),
  withTableSize(),
  withTranslation(),
)(MapPlayground);
