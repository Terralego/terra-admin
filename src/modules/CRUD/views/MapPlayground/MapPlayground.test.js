import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { toast } from '../../../../utils/toast';
import MapPlayground from './MapPlayground';

jest.mock('react-router-dom', () => ({
  Redirect: props => <div {...props}>Redirect</div>,
}));

jest.mock('@terralego/core/modules/Map/InteractiveMap', () => {
  const InteractiveMap = () => <div>InteractiveMap</div>;
  InteractiveMap.INTERACTION_FN = 'function';
  return InteractiveMap;
});

jest.mock('@terralego/core/modules/Map', () => ({}));

jest.mock('../../../../components/Message', () => () => <div>No settings</div>);

jest.mock('../../components/DataTable', () => () => <div>Datatable</div>);

jest.mock('../../components/DetailsWrapper', () => ({ children }) => children);

jest.mock('../../components/Details', () => () => <div>Details</div>);

jest.mock('./components/Map', () => () => <div>Map</div>);

jest.mock('../../services/CRUD', () => ({
  getView: (settings, layer) => ['layerTest', 'layerTest2'].includes(layer) && ({
    layer: {
      id: 1,
      tilejson: '/api/layer/1/tilejson',
      name: 'Layer test',
      geom_type: 1,
      settings: {},
    },
    name: 'Layer test',
    formSchema: {
      required: [
        'name',
      ],
      properties: {
        name: {
          type: 'string',
          title: 'Nom',
        },
        composante: {
          type: 'boolean',
          title: 'Composante',
        },
        numero: {
          type: 'integer',
          title: 'NumÃ©ro',
        },
      },
    },
    uiSchema: {},
    templates: [],
    extent: [1, 2, 3, 4],
  }),
  getFirstCrudViewName: () => 'layerTest',
  getSources: () => [{
    id: '1',
    type: 'vector',
    url: '/api/layer/1/tilejson',
  }, {
    id: '2',
    type: 'vector',
    url: '/api/layer/2/tilejson',
  }],
  getLayers: () => [{
    id: 'terralego-layerTest-1',
    'source-layer': 'layerTest',
    source: '1',
    type: 'line',
    paint: {
      'line-color': 'green',
      'line-width': 3,
    },
  }, {
    id: 'terralego-layerTest2-2',
    'source-layer': 'layerTest2',
    source: '2',
    type: 'line',
    paint: {
      'line-color': 'red',
      'line-width': 8,
    },
  }],
}));

jest.mock('../../../../utils/toast', () => ({
  toast: {
    displayError: jest.fn(),
  },
}));

jest.mock('../../../../components/Loading', () => () => <div>Loading</div>);

const settings = {
  menu: [{
    id: 1,
    crud_views: [{
      id: 1,
      name: 'Layer test',
      order: 0,
      map_style: {
        type: 'line',
        paint: {
          'line-color': 'green',
          'line-width': 3,
        },
      },
      layer: {
        id: 1,
        name: 'layerTest',
        tilejson: '/api/layer/1/tilejson',
      },
    }],
  }, {
    id: 2,
    crud_views: [{
      id: 2,
      name: 'Layer test 2',
      order: 0,
      map_style: {
        type: 'line',
        paint: {
          'line-color': 'red',
          'line-width': 8,
        },
      },
      layer: {
        id: 2,
        name: 'layerTest2',
        tilejson: '/api/layer/2/tilejson',
      },
    }],
  }],
  config: {
    default: {
      map: {
        accessToken: 'ACCESSTOKEN',
        center: [2, 46],
        zoom: 5,
        maxZoom: 18,
        minZoom: 3,
      },
    },
  },
};

// let props;
// beforeEach(() => {
const props = {
  featureToHighlight: jest.fn(),
  errors: {
    settings: undefined,
  },
  getSettings: jest.fn(),
  settings,
  settingsEndpoint: 'path/to/endpoint',
  t: key => key,
};
// });

describe('snapshots', () => {
  it('should display no settings message', () => {
    const tree = renderer.create((
      <MapPlayground
        {...props}
        errors={{
          settings: 'Not found',
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const tree = renderer.create((
      <MapPlayground
        {...props}
        match={{ params: { id: 'foo', layer: 'layerTest' } }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not display details when the feature id is not defined', () => {
    const tree = renderer.create((
      <MapPlayground
        {...props}
        match={{ params: { id: undefined, layer: 'layerTest' } }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display loading', () => {
    const tree = renderer.create((
      <MapPlayground
        {...props}
        settings={{}}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});


it('should active datatable when the layer is defined', () => {
  const wrapper = shallow((
    <MapPlayground
      {...props}
      match={{ params: { id: undefined, layer: 'layerTest' } }}
    />
  ));
  const table = wrapper.find('.CRUD-table');
  expect(table.hasClass('CRUD-table--active')).toBe(true);
});

it('should not active datatable when the layer and id are defined', () => {
  const wrapper = shallow((
    <MapPlayground
      {...props}
      match={{ params: { id: '1', layer: 'layerTest' } }}
    />
  ));
  const table = wrapper.find('.CRUD-table');
  expect(table.hasClass('CRUD-table--active')).toBe(false);
});

it('should redirect to the root view when trying to access to an unknown layer', () => {
  shallow((
    <MapPlayground
      {...props}
      match={{ params: { id: '1', layer: 'notExists' } }}
    />
  ));
  expect(toast.displayError).toHaveBeenCalled();
});

it('should redirect to the first layer', () => {
  const tree = renderer.create((
    <MapPlayground
      {...props}
      match={{ params: { id: '1', layer: undefined } }}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});


it('should else get settings when mounting', () => {
  const instance = new MapPlayground({
    ...props,
    settings: undefined,
    match: { params: { id: '1', layer: undefined } },
  });

  instance.componentDidMount();
  expect(instance.props.getSettings).toHaveBeenCalled();
});


describe('add or remove highlight on table hover cell', () => {
  const instance = new MapPlayground({
    ...props,
    match: { params: { layer: 'layerTest' } },
    map: { foo: 'foo' },
  });
  instance.state = {
    customStyle: {
      layers: [{
        id: 1,
        'source-layer': 'layerTest',
      }, {
        id: 2,
        'source-layer': 'layerTest2',
      }, {
        id: 3,
        'source-layer': 'notExistsInTheMap',
      }],
    },
  };

  it('should highlight feature on table hover cell', () => {
    instance.onTableHoverCell('1');
    expect(instance.props.featureToHighlight).toHaveBeenCalledWith({
      featureId: '1',
      layer: 'layerTest',
      hover: true,
    });
  });

  it('should remove highlight feature on table hover cell', () => {
    jest.clearAllMocks();
    instance.onTableHoverCell('2', false);
    expect(instance.props.featureToHighlight).toHaveBeenCalledWith({
      featureId: '2',
      layer: 'layerTest',
      hover: false,
    });
  });
});
