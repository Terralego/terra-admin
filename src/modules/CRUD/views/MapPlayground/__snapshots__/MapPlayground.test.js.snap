// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should redirect to the first layer 1`] = `
<div
  to="/CRUD/map/layerTest"
>
  Redirect
</div>
`;

exports[`snapshots should display loading 1`] = `
<div
  to="/CRUD/map/layerTest"
>
  Redirect
</div>
`;

exports[`snapshots should display no settings message 1`] = `
<div>
  No settings
</div>
`;

exports[`snapshots should not display details when the feature id is not defined 1`] = `
Array [
  <div
    className="CRUD-map CRUD-map--with-table-medium"
  >
    <div>
      Map
    </div>
  </div>,
  <div
    className="CRUD-table CRUD-table--active CRUD-table--medium"
    onTransitionEnd={[Function]}
  >
    <div>
      Datatable
    </div>
  </div>,
]
`;

exports[`snapshots should render correctly 1`] = `
Array [
  <div>
    Details
  </div>,
  <div
    className="CRUD-map"
  >
    <div>
      Map
    </div>
  </div>,
  <div
    className="CRUD-table"
    onTransitionEnd={[Function]}
  >
    <div>
      Datatable
    </div>
  </div>,
]
`;
